xline/src/server/watch_server.rs

```diff
@@ -0,0 +1,263 @@
+    request_rx: tonic::Streaming<WatchRequest>,
+    /// Event receiver
+    event_rx: mpsc::Receiver<WatchEvent>,
+    /// Event sender
+    event_tx: mpsc::Sender<WatchEvent>,
+    /// Watch ID to watcher map
+    watcher_map: HashMap<WatchId, Watcher>,
+    /// Next available `WatchId`
+    next_id: WatchId,
+}
+
+impl WatchHandle {
+    /// New `WatchHandle`
+    fn new(
+        kv_watcher: Arc<KvWatcher>,
+        response_tx: mpsc::Sender<Result<WatchResponse, tonic::Status>>,
+        request_rx: tonic::Streaming<WatchRequest>,
+        event_rx: mpsc::Receiver<WatchEvent>,
+        event_tx: mpsc::Sender<WatchEvent>,
+    ) -> Self {
+        Self {
+            kv_watcher,
+            response_tx,
+            request_rx,
+            event_rx,
+            event_tx,
+            watcher_map: HashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    /// Get next available `WatchId`, return -1 if id is not available
+    fn get_next_watch_id(&mut self, id: WatchId) -> WatchId {
+        // 0 means auto-generate
+        if id == 0 {
+            loop {
+                let next = self.next_id;
+                self.next_id = self.next_id.overflow_add(1);
+                if self.watcher_map.get(&next).is_none() {
+                    break next;
+                }
+            }
+        } else if self.watcher_map.get(&id).is_some() {
+            -1
+        } else {
+            id
+        }
+    }
+
+    /// Handle `WatchRequest`
+    async fn handle_watch_request(&mut self, req: WatchRequest) {
+        if let Some(req) = req.request_union {
+            match req {
+                RequestUnion::CreateRequest(req) => {
+                    let watch_id = self.get_next_watch_id(req.watch_id);
+                    if watch_id == -1 {
+                        assert!(
+                            self.response_tx
+                                .send(Err(tonic::Status::already_exists(format!(
+                                    "Watch ID {} has already been used",
+                                    req.watch_id
+                                ))))
+                                .await
+                                .is_ok(),
+                            "Receiver is closed"
+                        );
+                        return;
+                    }
+
+                    let key_range = KeyRange {
+                        start: req.key,
+                        end: req.range_end,
+                    };
+
+                    let (watcher, events, revision) = self
+                        .kv_watcher
+                        .watch(
+                            watch_id,
+                            key_range,
+                            req.start_revision,
+                            req.filters,
+                            self.event_tx.clone(),
+                        )
+                        .await;
+                    let _prev = self.watcher_map.insert(watch_id, watcher);
```

Comment:
the prev should be none, we'd better make sure it's none, or comment the reason to ignore the return value.

Reply:
Done